Create main.cpp

MaEgV opened the pull request.

MaEgV committed:
Create main.cpp

MaEgV committed:
Add files via upload

MaEgV commented:
Структура наследования:
object -> borders
           -> block
           -> moving_objects -> ball, racket, bonus
Интерфейсный класс для всего уровня - класс "level", в отдельном файле
Объекты спрятаны в стеке и могут взаимодействовать только с помощью перегруженного метода action(). Инициируют взаимодействия движущиеся объекты, поэтому action() перегружен только для них
Для каждого двигающегося объекта стек возвращает все объекты, с которыми произошло столкновение, и для каждого столкновения вызывается action(this). Таким образом можно обработать все виды взаимодействий в игре

Gandy123 commented on game_objects.h:
```diff
@@ -0,0 +1,150 @@
+using namespace std;
+
+int lives;
+
```
глобальная переменная в h-нике плохо. Нарушение инкапсуляции и прочее

MaEgV replied:
c-шников пока нет, поэтому она здесь)

Gandy123 commented on game_objects.h:
```diff
@@ -0,0 +1,150 @@
+
+public:
+	object_stack();  // something from @level@
+
```
вот здесь бы поподробнее как вы собираетесь это стек заполнять например

MaEgV replied:
Думаю, файл с уровнем - это интерфейсная часть, поэтому парсится он должен в ней(level.h)
В стэк будет передаваться конкретный набор объектов, или, например, описание этого набора, чтоб сами объекты создавались внутри модуля с объектами. Мне кажется, это будет не стэк, а скорее список
Возможный вариант заполнения:
контейнер[0] - ракетка
контейнер[1] - мяч
контейнер[2] - граница поля
контейнер[3+] - блоки
Блоки должны быть упорядочены так, чтобы к нижним можно было попасть быстрее. Вероятность врезаться в них выше, чем у верхних
Этот же класс должен инициировать проверку столкновений между объектами, которые хранит

Gandy123 commented on game_objects.h:
```diff
@@ -0,0 +1,150 @@
+class bonus : public moving_objects {
+private:
+	void collision(object& a) override { a.action(this); }
+
```
virtual принято писать при наследовании и переопределении виртуальной функции

MaEgV replied:
Для того чтобы затем была возможность переопределить их в возможных потомках? Или просто так принято?)

Gandy123 replied:
так принято

Gandy123 commented on game_objects.h:
```diff
@@ -0,0 +1,150 @@
+public:
+	// different interactions between objects
+	virtual void action(ball* a) = 0;
+	virtual void action(racket* a) = 0;
```
странная тема когда в базовом объекте есть метод который завязан на наследников - как-то это .... давайте подумаем как убрать

MaEgV replied:
См. нижний апдейт

Gandy123 reviewed (changes requested):
done

MaEgV committed:
Второй вариант

Для обработки взаимодействия между объектами в корневой класс добавлено поле типа объекта
